prusti-encoder/src/encoders/type/typ.rs

```diff
@@ -0,0 +1,1373 @@
+                    snap_to_disc_snap,
+                    variants: vcx.alloc(variants),
+                }),
+                method_assign,
+                ref_to_pred: Self::mk_predicate_ident(name_p),
+                ref_to_snap,
+                unreachable_to_snap,
+            },
+        );
+
+        let mut funcs: Vec<vir::DomainFunction<'vir>> = vec![];
+        let mut axioms: Vec<vir::DomainAxiom<'vir>> = vec![];
+        let mut field_projection_p = Vec::new();
+        let mut other_predicates = Vec::new();
+
+        funcs.push(vcx.alloc(vir::DomainFunctionData {
+            unique: false,
+            name: s_discr_func_name,
+            args: vcx.alloc_slice(&[ty_s]),
+            ret: discr_type.snapshot,
+        }));
+
+        let mut predicate_accumulator = vcx.mk_bool::<false>();
+
+        let self_local = vcx.mk_local_ex("self_p");
+
+        let discr_field_access = vcx.mk_field_expr(self_local, field_discriminant);
+        let mut snap_accumulator = unreachable_to_snap.apply(vcx, []);
+
+        for (vidx, discr) in adt.discriminants(vcx.tcx) {
+            let variant = adt.variant(vidx);
+            let name_s =
+                vir::vir_format!(vcx, "s_Adt_{did_name}_{vidx:?}_{}", variant.name.as_str());
+            let name_p =
+                vir::vir_format!(vcx, "p_Adt_{did_name}_{vidx:?}_{}", variant.name.as_str());
+
+            let ref_to_pred = Self::mk_predicate_ident(name_p);
+            let discr_ty = deps.require_ref::<crate::encoders::TypeEncoder>(discr.ty).unwrap();
+            let idx_expr = discr_ty.expr_from_u128(discr.val);
+
+            let (_, cons_call) = Self::mk_enum_variant(
+                vcx,
+                deps,
+                variant,
+                vidx,
+                idx_expr,
+                ty_s,
+                name_s,
+                name_p,
+                ref_to_pred,
+                snap_to_disc_snap,
+                &mut funcs,
+                &mut field_projection_p,
+                &mut axioms,
+                &mut other_predicates,
+            );
+
+            let cond = vcx.mk_bin_op_expr(vir::BinOpKind::CmpEq, discr_field_access, idx_expr);
+
+            let pred_call = vcx.alloc(vir::ExprData::PredicateApp(ref_to_pred.apply(vcx, [self_local])));
+            predicate_accumulator =
+                vcx.mk_ternary_expr(cond, pred_call, predicate_accumulator);
+
+            snap_accumulator = vcx.mk_ternary_expr(cond, cons_call, snap_accumulator)
+        }
+
+        let variants_len = vcx.mk_const_expr(vir::ConstData::Int(adt.variants().len() as u128));
+
+        let predicate = {
+            let acc = vcx.mk_acc_field_expr(self_local, field_discriminant);
+            let discr_field_access_int = discr_type.expect_prim().snap_to_prim.apply(vcx, [discr_field_access]);
+
+            // TODO: handle the empty enum? i guess the lower and upper bound together for an empty enum are false which is correct?
```

Comment:
We could make this a special case to spell out the `false` in Viper code, but yes, I would say that the predicate for `enum Empty {}` is `predicate p_Empty(..) { false }`. Same for any "uninhabited" type (see [`is_inhabited_from`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/struct.Ty.html#method.is_inhabited_from)), including the never type `!`. However:

- Any method which accepts a parameter of type `Empty` cannot be executed, i.e. its precondition should be `false`. However, Viper would only see this if we actually unfolded the predicate.
- Any method which "returns" a value of type `Empty` cannot terminate, i.e. its postcondition should be `false`. Again, this would only be seen if we unfolded the predicate.
- Any pure *function* which "returns" a value of type `Empty` should be rejected. It's probably best to do this pretty early on, to avoid accidentally emitting a function whose definitional axiom allows us to assume `false`. Maybe checking that the return type of pure functions is inhabited is sufficient.

We can worry about this in a separate PR though.